### DISCUSSION TITLE
Give user more flymake control

### DISCUSSION THREAD

**Discussion by Dickby** [joaotavora/eglot#964] [General]
eglot seems to totally ignore the state of flymake-mode, disabling it changes nothing inside the eglot-managed buffer.
Further the varibale `flymake-no-changes-timeout` is a floating point number,
but it seems eglot just uses it like a boolean, setting it to nil disables automatic checking,
every other value just acts as `t`.

**Comment by joaotavora** [comment]
See this commit message and related issues for a discussion of why `flymake-no-changes-timeout` cannot be honoured in the traditional way: https://github.com/joaotavora/eglot/commit/2b87b06d9ef15e7c39d87fd5a4375b6deaa7e322

> eglot seems to totally ignore the state of flymake-mode, disabling it changes nothing inside the eglot-managed buffer.

What are you trying to achieve?  Do you want to not use flymake?  Then see variable `eglot-stay-out-of`.

**Reply by joaotavora** [reply]
See also: https://github.com/joaotavora/eglot/pull/957#issuecomment-1131420325

**Reply by Dickby** [reply]
I use eglot with clangd as server.
I try to find a setting so that flymake is not invoked while i type.
Setting `flymake-no-changes-timeout` to `nil` and `flymake-start-on-save-buffer` to `t` doesn't work properly.
Often if i save the buffer flymake highlights errors that are from an earlier state of the buffer.
To  make flymake recheck, i have to change something in the buffer and save it again.
Another way that would work for me is to only check the buffer after an idle time, and i thought that's what `flymake-no-changes-timeout` is for.
Maybe i have to set it to nil and than run `flymake-start` with an idle timer myself. 
But shouldn't eglot provide such an option?
Thanks

**Reply by joaotavora** [reply]
The comment I linked to explains why what you're trying to do is currently impossible with LSP (as far as I know).  The LSP spec does not specify a way to request diagnostics "on demand". This is an essential primitive operation to get the behaviour you're looking for.  Instead, an LSP server provides them "when it sees fit": usually -- but not always -- after it is informed of changes in the file.   So unless the LSP spec has changed in that regard (it might have, maybe someone knows), I'm afraid your use case cannot be implemented.

**Reply by Dickby** [reply]
Thanks for the information.
I try to hack around that then.

**Reply by joaotavora** [reply]
Let us know how it goes.

**Reply by Dickby** [reply]
if i set `flymake-no-changes-timeout` to `nil`, is flymake still using eglot's informations for the check if i run `flymake-start` ?

**Reply by joaotavora** [reply]
Yes.  That's what the commit message I linked to says.

**Comment by Dickby** [comment]
Sorry i didn't understand it from that, my bad.
But in that case this is what works for me:
```
  (setq flymake-no-changes-timeout nil)
  (add-hook 'c-mode-hook (lambda()
			   (setq-local my/flymake-idle (run-with-idle-timer 2 t 'flymake-start))
			   (add-hook 'kill-buffer-hook (lambda()
							 (when (timerp my/flymake-idle )
							   (cancel-timer my/flymake-idle))) 0 t)))
```

**Reply by joaotavora** [reply]
Alright, as I wrote in #957, you may be lucky with that strategy or you may be not.  If the LSP server takes, say, 2.1 seconds to deliver diagnostics to the client after being informed of changed, then you'll always be one step behind. And that time depends on the project size etc.  So I don't think this is a a general solution.  But if it works for you and your project, great!

**Reply by Dickby** [reply]
Isn't there a way to tell when the last time the server delivered diagnostics, and run flymake only if that has happend during the idle time?

**Reply by Dickby** [reply]
Now that i tried a bit: my solution is not optimal either.
It's nice that i can control when flycheck gets the results of the changes presented to the lsp-server.
But I think for custom tweaking it would be really good to be able to control when eglot sends the changes to the server.
Depending on the users configuration, there could be nice ways to set it up.
For evil-mode for example it could nicely be controled by the evil-state changes.

**Comment by Dickby** [comment]
I see it is mentioned there too. 
Thanks 
feel free to close this issue
